kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala

```diff
@@ -143,7 +147,21 @@ class BatchJobSubmission(
     val asyncOperation: Runnable = () => {
       setStateIfNotCanceled(OperationState.RUNNING)
       try {
-        submitBatchJob()
+        // If the batch submission is in recovery mode, try to get submitted applicationId
+        // from metadata or resource manager application status. If got, monitor the submitted
+        // application. Otherwise, resubmit the batch job.
+        val submittedAppId = recoveryMetadata.map { metadata =>
```

Comment:
We can make here more clear that, we only try to re-submit engine if previous status is pending and we failed to fetch the status from resource manager. And if the application status is terminated, we need to update the session store state.

```
previous is pending -> re-submit -> same code path with normal
previous is running -> check current state -> still monitor if running
                                           -> update store state if terminated
```